### PR TITLE
Display corrections in context menu for #54

### PR DESCRIPTION
**Pull request by speter52**
Displaying the corrections for a misspelled word as context menu options as described in #54. This is my first OSS contribution so any feedback on the design or workflow would be great.

### CODE REVIEW

File: lib/spell-check-view.coffee [hunk 1]
```diff
@@ -107,3 +111,76 @@ class SpellCheckView
...
+    # Listen for click events from other editors and clear the context menu entries if
+    # this view's editor isn't the active editor.
+    atom.document.addEventListener 'mousedown', =>
+      if @editor.id isnt atom.workspace.getActiveTextEditor().id
+        @clearContextMenuEntries()
+
+  addContextMenuEntries: (mouseEvent) =>
+    @clearContextMenuEntries()
+    # Check to see if a context menu was opened through a right mouse click.
+    if mouseEvent.button is 2
```
Comment (50Wliu) [inline lib/spell-check-view.coffee]: Why do we care how the menu was opened?

File: lib/spell-check-view.coffee [hunk 2]
```diff
@@ -107,3 +111,76 @@ class SpellCheckView
     instance = @getInstance()
     misspelling = @editor.getTextInBufferRange marker.getBufferRange()
     instance.suggest args, misspelling
+
+  initializeContextMenuListeners: =>
+    atom.views.getView(@editor).addEventListener 'mousedown', @addContextMenuEntries
```
Comment (50Wliu) [inline lib/spell-check-view.coffee]: You'll need to find a better way to detect when a context menu is opened - there are ways to open it without using the mouse (for example, Windows has a dedicated keyboard key that opens the menu).

Comment (speter52) [inline lib/spell-check-view.coffee]: Ah right. I assumed that context menus were only opened with right-mouse clicks so thought a `mousedown` listener would work. I'll look into a more reliable way of detecting when context menus open/close.

**Comment by speter52** [comment]
I found a `contextmenu` event in the Electron code which fires when context menus open, so I updated the listener to watch for that. I also modified the `SpellCheckView` class to hold a reference to a list of all context menu items that are added to the global registry - which is shared between the `SpellCheckView`'s of other editor instances. This way the menu items can be disposed as needed. I had some difficulty properly sharing that list between different `SpellCheckView`'s which I elaborated on at line 64 in `lib/main.coffee`.

Also the Travis CI check seems to be taking a lot longer than before?

**Comment by speter52** [comment]
Thanks, what would the next steps be?

**Comment by 50Wliu** [comment]
For someone with proper experience with the spell-check package to give this a review.  I believe that would be @as-cii.

**Comment by as-cii** [comment]
This looks awesome, @speter52! I found a little quirk when testing this out: when a word has no corrections, `No corrections` is shown as expected. When closing the context menu and reopening it, `No corrections` is shown without a separator, and never goes away from there on.

context-menu

Honestly I think we might get rid of the `No corrections` menu item, as it doesn't perform any action when clicked.

Other than this little nuance, I think this is ready to 🚢, so I will hit the merge button after addressing the minor problem described above. Nice work @speter52! 💯

**Comment by speter52** [comment]
Thanks a lot @as-cii! I just fixed the 'No corrections' issue. I had missed that use case when refactoring how the TextEditor's stored the context menu entries. Let me know if you see anything else needed from my end.

**Comment by 50Wliu** [comment]
I agree that we should get rid of the "No corrections" menu, if just to stay consistent with how Chrome does it.

File: lib/spell-check-view.coffee [hunk 3]
```diff
@@ -107,3 +110,64 @@ class SpellCheckView
...
+            'atom-text-editor': [{label: correction.label, command: 'spell-check:correct-misspelling-' + correction.index}]
+          })
+          @spellCheckModule.contextMenuEntries.push contextMenuEntry
+
+  makeCorrection: (correction, marker) =>
+    if correction.isSuggestion
+      # Update the buffer with the correction.
+      @editor.setSelectedBufferRange(marker.getBufferRange())
+      @editor.insertText(correction.suggestion)
+    else
```
Comment (as-cii) [inline lib/spell-check-view.coffee]: Do we need to handle the `else` scenario? If so, can we add a test for it?

**Comment by as-cii** [comment]
I made a couple of changes and left an additional comment on a line of code that I missed during the first review. Other than that, this looks ready to merge! Again, great work! 💯

File: lib/spell-check-view.coffee (hunk 3 above)
Comment (speter52) [inline lib/spell-check-view.coffee]: I had brought this use case over from the `CorrectionsView:confirm()` function to preserve the 'Add to Known Words' functionality, it looks like that's the only way it's being used right now. I added a spec now to test adding a known word from the context menu.

**Comment by speter52** [comment]
Just pushed the review changes!

**Comment by as-cii** [comment]
Awesome, thanks @speter52! 💥

**Comment by cameronmcefee** [comment]
> This is my first OSS contribution so any feedback on the design or workflow would be great.

👏 👏 👏  You're off to a great start. Thanks so much for this.

**Comment by jbjonesjr** [comment]
:+1: this is great news. Would use your contributions again

**Comment by colinbrislawn** [comment]
Well done! This program has needed this feature for ages!

**Comment by speter52** [comment]
Thanks for the comments, happy to contribute! 😀

**Pull request merged**